### ISSUE TITLE
This repo shouldn't exist

### ISSUE THREAD

**Issue by mvf** [surge-synthesizer/surge-filesystem#7]
Having only now looked into how this code is intended to be (re-)used, **I strongly suggest removing this repo and just using C++17 in the other project, [ShortCircuit](https://github.com/surge-synthesizer/shortcircuit3)**.

Intuitively, moving code with a clean interface to its own repo for reuse makes sense. However, **this particular implementation solves a temporary problem specific to the Surge synth**. Proliferating it into new projects creates technical debt without benefit.

### TL;DR

- The sole purpose of this code is for the Surge synth to keep supporting old macOS.
- Supporting old macOS is a pain, and `filesystem` is just a small part of that pain.
- ShortCircuit is Windows-only today.
- An eventual macOS port of ShortCircuit doesn't need to run on old macOS.

#### The purpose of this code

The Surge synth is available for macOS, and it uses `std::filesystem` extensively. `std::filesystem` support is only available in fairly recent versions of macOS and linking `libc++` statically is unsupported. So we implemented this to not leave users on old macOS versions behind.

This is not a complete filesystem implementation, and what is there is intentionally not 100% compatible with `std::filesystem` in order to help with the safety of UTF-8 paths Surge uses on Windows. **This approach is a transitional hack in the Surge synth and will be removed in Surge 1.9.** ShortCircuit should be ported straight from the Windows API to `std::filesystem` (most importantly the `path` class). This will make the resulting code much cleaner.

#### The cost of supporting legacy macOS

Supporting old macOS versions is a burden. Most importantly, `std::filesystem` is a C++17 API, but the old macOS'en are stuck on C++14. That's why Surge today is C++14 on macOS and C++17 everywhere else. This is not a great situation for devs, as they might find their portable code doesn't compile for macOS.

Legacy macOS support will be dropped from Surge at some point not too far into the future. Let's not encumber a new project with this unholy split between C++14 and 17.

#### ShortCircuit is a Windows program

The only other intended user of this code, ShortCircuit, is a pure Windows program. There are no existing macOS users, so it should just be made pure C++17. In fact, it would be prudent to keep ShortCircuit Windows-only for now until it's fully working (VST3, etc.). Once that's done we're in a much better position to judge whether cross-platform is even feasible. Porting a working program is a lot easier too.

**Comment by baconpaul** [comment]
Yeah that’s not the right answer

1. macOS 10.15 is the first macOS to support std::filesystem
2. Gcc9 is the first gcc to support it
3. ShortCircuit3 right now compiles on macos and gcc older than those versions
4. Many many many of our users use those versions

Also ShortCircuit isn’t a pure windows program any more. I’ve spent the past couple of weeks making it work on mac and linux, and it plays samples there and works right now. The windows specific APIs have been replaced or ifdefed around with portable / posix versions. Azure pipelines and regtests work all thre platforms today.

If we don’t want this repo to exist, I could bind both SC and Surge to https://github.com/gulrak/filesystem Instead I suppose.

**Comment by baconpaul** [comment]
Oh bug I agree we should use fs::path everywhere of course. If you look at the recent commits in SC3 it has been doing that as much as possible.

So should we dump this and use that other implementation of filesystem on macos and linux before gcc9? Fine with me.

**Comment by baconpaul** [comment]
So over in `https://github.com/surge-synthesizer/shortcircuit3/pull/61` I move SC3 to that gulrak filesystem 
it seems to be the solution. Also (understandably) is more complete than our hack here.
I have a few errands and meetings this morning but I might merge that (or merge a variation on that) tonight and then will delete this repo and still support macOS 10.12 and higher in SC3 and gcc 7 and higher in linux!

**Comment by baconpaul** [comment]
OK so that ghc filesystem works great
I just copied a teensy bit of the cmake detection rules and added an "infrastructure/import_fs.h" header to SC3 (which includes path_to_strign and string_to_path) and everything works.
So I'll merge that and delete this repo.

**Comment by baconpaul** [comment]
As of 264c887 this repo is now empty at head.

I am leaving it around for a week or two while people update their SC3 to the version which doesn't need it thenI will delete the repo altogether.

**Comment by mvf** [comment]
I'm fine with this solution, even if it's not exactly what I had in mind. Keeping the filesystem in the Surge repo will make it slightly easier to work with after all. But starting a new project on a C++14/C++17 mix still feels like a missed opportunity.

> ShortCircuit3 right now compiles on macos and gcc older than those versions

Gotcha, but this is only as of a few weeks ago, right? My point was that no existing users will be left behind by going C++17.

> Many many many of our users use those versions

In the same vein, right now ShortCircuit3 has zero users on these old versions because it has only been released on Windows.

IMO a codebase that is C++14 only on macOS and C++17 elsewhere makes it harder to deliver value for *all* users.

But, ultimately, whoever does the work gets to call the shots, and for ShortCircuit3 that's not me. 😉

**Comment by baconpaul** [comment]
I actually think we should take the same approach (17 with the other filesystem) in surge also. We can get to 17 with macos 10.12 if we use a fs impl and that one I picked for sc3 looks good. I wrote it up in a surge issue. 

The code we got for sc didn’t make a working windows plugin either so we are basically doing a rewrite of big chunks. I can walk you through some of our choices if you want! But since it is a redo let’s do it cross platform. And support the footprint we saw our surge users use. That’s basically the conclusion.

Thanks for pushing though. Better outcome
